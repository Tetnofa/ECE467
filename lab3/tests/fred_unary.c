void f1() {
  int x;
  int y;
  bool b1;
  bool b2;
  
  -5;
  -x;
  -(-10);
  -(y + 3);

  !true;
  !b1;
  !(false || true);
  !(!b2);
  !(b1 && b2);
  !((b1 || false) && (b2 || true));

  !5;          // Invalid: '!' applied to int instead of bool
  -x &&y;      // Invalid: '-' should apply to entire integer expression
  -(true);     // Invalid: '-' applied to bool instead of int
  -(!b1);      // Invalid: '-' applied to result of boolean expression
  -true;       // Invalid: '-' applied to a bool constant
  !(x + y);    // Invalid: '!' applied to int expression
  !(-x);       // Invalid: '!' applied to result of integer expression
  -(b1 || b2); // Invalid: '-' applied to boolean expression
}
