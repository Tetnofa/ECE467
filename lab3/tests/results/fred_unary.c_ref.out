Program [useIO=0](1,0)
	Function[isProto=0]	Void(1,0)
		Identifier[name:f1](1,5)
		Scope(1,10)
			Scalar Declaration			Int(2,2)
				Identifier[name:x](2,6)
			Scalar Declaration			Int(3,2)
				Identifier[name:y](3,6)
			Scalar Declaration			Bool(4,2)
				Identifier[name:b1](4,7)
			Scalar Declaration			Bool(5,2)
				Identifier[name:b2](5,7)
			Expression Statement(7,2)
				Int Expression(7,2)
					IntConstant[val=-5](7,2)
			Expression Statement(8,2)
				UnaryExpr[opcode:-](8,2)
					Int Expression(8,3)
						Reference(8,3)
							Identifier[name:x](8,3)
			Expression Statement(9,2)
				UnaryExpr[opcode:-](9,2)
					Int Expression(9,4)
						IntConstant[val=-10](9,4)
			Expression Statement(10,2)
				UnaryExpr[opcode:-](10,2)
					Int Expression(10,4)
						Binary Expression[opcode:+](10,4)
							Int Expression(10,4)
								Reference(10,4)
									Identifier[name:y](10,4)
							Int Expression(10,8)
								IntConstant[val=3](10,8)
			Expression Statement(12,2)
				UnaryExpr[opcode:!](12,2)
					Int Expression(12,3)
						BoolConstant[val=1](12,3)
			Expression Statement(13,2)
				UnaryExpr[opcode:!](13,2)
					Int Expression(13,3)
						Reference(13,3)
							Identifier[name:b1](13,3)
			Expression Statement(14,2)
				UnaryExpr[opcode:!](14,2)
					Bool Expression(14,4)
						Binary Expression[opcode:||](14,4)
							Int Expression(14,4)
								BoolConstant[val=0](14,4)
							Int Expression(14,13)
								BoolConstant[val=1](14,13)
			Expression Statement(15,2)
				UnaryExpr[opcode:!](15,2)
					UnaryExpr[opcode:!](15,4)
						Int Expression(15,5)
							Reference(15,5)
								Identifier[name:b2](15,5)
			Expression Statement(16,2)
				UnaryExpr[opcode:!](16,2)
					Bool Expression(16,4)
						Binary Expression[opcode:&&](16,4)
							Int Expression(16,4)
								Reference(16,4)
									Identifier[name:b1](16,4)
							Int Expression(16,10)
								Reference(16,10)
									Identifier[name:b2](16,10)
			Expression Statement(17,2)
				UnaryExpr[opcode:!](17,2)
					Bool Expression(17,4)
						Binary Expression[opcode:&&](17,4)
							Bool Expression(17,5)
								Binary Expression[opcode:||](17,5)
									Int Expression(17,5)
										Reference(17,5)
											Identifier[name:b1](17,5)
									Int Expression(17,11)
										BoolConstant[val=0](17,11)
							Bool Expression(17,22)
								Binary Expression[opcode:||](17,22)
									Int Expression(17,22)
										Reference(17,22)
											Identifier[name:b2](17,22)
									Int Expression(17,28)
										BoolConstant[val=1](17,28)
			Expression Statement(19,2)
				UnaryExpr[opcode:!](19,2)
					Int Expression(19,3)
						IntConstant[val=5](19,3)
			Expression Statement(20,2)
				Bool Expression(20,2)
					Binary Expression[opcode:&&](20,2)
						UnaryExpr[opcode:-](20,2)
							Int Expression(20,3)
								Reference(20,3)
									Identifier[name:x](20,3)
						Int Expression(20,7)
							Reference(20,7)
								Identifier[name:y](20,7)
			Expression Statement(21,2)
				UnaryExpr[opcode:-](21,2)
					Int Expression(21,4)
						BoolConstant[val=1](21,4)
			Expression Statement(22,2)
				UnaryExpr[opcode:-](22,2)
					UnaryExpr[opcode:!](22,4)
						Int Expression(22,5)
							Reference(22,5)
								Identifier[name:b1](22,5)
			Expression Statement(23,2)
				UnaryExpr[opcode:-](23,2)
					Int Expression(23,3)
						BoolConstant[val=1](23,3)
			Expression Statement(24,2)
				UnaryExpr[opcode:!](24,2)
					Int Expression(24,4)
						Binary Expression[opcode:+](24,4)
							Int Expression(24,4)
								Reference(24,4)
									Identifier[name:x](24,4)
							Int Expression(24,8)
								Reference(24,8)
									Identifier[name:y](24,8)
			Expression Statement(25,2)
				UnaryExpr[opcode:!](25,2)
					UnaryExpr[opcode:-](25,4)
						Int Expression(25,5)
							Reference(25,5)
								Identifier[name:x](25,5)
			Expression Statement(26,2)
				UnaryExpr[opcode:-](26,2)
					Bool Expression(26,4)
						Binary Expression[opcode:||](26,4)
							Int Expression(26,4)
								Reference(26,4)
									Identifier[name:b1](26,4)
							Int Expression(26,10)
								Reference(26,10)
									Identifier[name:b2](26,10)
sema: 19:3 : type mismatch 
sema: 20:2 : type mismatch 
sema: 21:4 : type mismatch 
sema: 22:4 : type mismatch 
sema: 23:3 : type mismatch 
sema: 24:4 : type mismatch 
sema: 25:4 : type mismatch 
sema: 26:4 : type mismatch 
